src/main/java/org/elasticsearch/common/compress/CompressedString.java

```diff
@@ -120,14 +102,19 @@ public boolean equals(Object o) {
 
         CompressedString that = (CompressedString) o;
 
-        if (!Arrays.equals(bytes, that.bytes)) return false;
+        if (Arrays.equals(compressed(), that.compressed())) {
+            return true;
+        }
 
-        return true;
+        return Arrays.equals(uncompressed(), that.uncompressed());
     }
 
     @Override
     public int hashCode() {
-        return bytes != null ? Arrays.hashCode(bytes) : 0;
+        if (hashCode == 0) {
```

Comment:
This is correct. Would you like it better if I did it this way so that the hashcode is never recomputed twice?

```java
if (hashcode == 0) {
  int h = Arrays.hashCode(uncompressed());
  if (h == 0) {
    h = 1
  }
  this.hashcode = h;
}
```